core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+        synchronized (banedParents) {
+            if (lastDiscoveryNodes == null || lastDiscoveryNodes.nodeExists(parentNode)) {
+                // Only set the ban if the node is the part of the cluster
+                banedParents.put(ban, reason);
+            }
+        }
+
+        // Now go through already running tasks and cancel them
+        for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+            Task task = taskEntry.getValue();
+            if (parentId == task.getParentId() && parentNode.equals(task.getParentNode())) {
+                assert task instanceof CancellableTask;
+                try {
+                    ((CancellableTask)task).cancel(reason);
+                } catch (Throwable ex) {
+                    logger.warn("Failed to cancel task with id {}", ex, task.getId());
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes the ban for the specified parent task.
+     *
+     * This method is called when a previously banned task finally cancelled
+     */
+    public void removeBan(String parentNode, long parentId) {
+        banedParents.remove(new Tuple<>(parentNode, parentId));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (lastDiscoveryNodes == null || event.nodesRemoved())  {
+            synchronized (banedParents) {
+                lastDiscoveryNodes = event.state().getNodes();
+                if (banedParents.isEmpty() == false) {
+                    // Remove all bans that were registered by nodes that are no longer in the cluster state
+                    Iterator<Tuple<String, Long>> banIterator = banedParents.keySet().iterator();
+                    while (banIterator.hasNext()) {
+                        Tuple<String, Long> nodeAndTaskId = banIterator.next();
+                        if (lastDiscoveryNodes.nodeExists(nodeAndTaskId.v1()) == false) {
+                            logger.debug("Removing ban for the parent [{}:{}] on the node [{}], reason: the parent node is gone", nodeAndTaskId.v1(), nodeAndTaskId.v2(), event.state().getNodes().localNode());
+                            banIterator.remove();
+                        }
+                    }
+                }
+            }
+            // Cancel cancellable tasks for the nodes that are gone
+            for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+                Task task = taskEntry.getValue();
+                String parent = task.getParentNode();
+                if (parent != null && lastDiscoveryNodes.nodeExists(parent) == false && task instanceof CancellableTask) {
+                    CancellableTask cancellableTask = (CancellableTask) task;
+                    if (cancellableTask.cancelOnParentLeaving()) {
+                        try {
+                            cancellableTask.cancel("Coordinating node [" + parent + "] left the cluster");
+                        } catch (Throwable ex) {
```

Comment:
either we let it blow up or we suppress the exception but I think logging is wrong here there should not be an exception?